File: AOS/app/src/test/java/com/example/escapephone/PlaytestAnalyticsTests.kt
package com.example.escapephone

import com.example.escapephone.app.GameViewModel
import com.example.escapephone.core.model.PuzzleExitReason
import com.example.escapephone.core.motion.MockMotionController
import com.example.escapephone.core.persistence.InMemoryGameProgressStore
import com.example.escapephone.core.persistence.TimeProvider
import com.example.escapephone.core.services.NoOpAdGateway
import com.example.escapephone.core.services.NoOpHapticProvider
import com.example.escapephone.core.services.NoOpPuzzleDeviceConnector
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

class PlaytestAnalyticsTests {
    private class MutableTimeProvider(var current: Long) : TimeProvider { override fun now() = current }
    private fun makeViewModel(store: InMemoryGameProgressStore, time: MutableTimeProvider, consent: Boolean = true) = GameViewModel(store, MockMotionController(), NoOpHapticProvider, NoOpAdGateway, NoOpPuzzleDeviceConnector(), time).also { if (consent) it.grantAnalyticsConsent() }

    @Test fun completePuzzleSession_savesElapsedTime() {
        val store = InMemoryGameProgressStore(); val time = MutableTimeProvider(1_000); val viewModel = makeViewModel(store, time)
        viewModel.startPuzzleSession("encrypted_note"); time.current = 4_500; viewModel.completePuzzleSession("encrypted_note"); viewModel.save()
        assertEquals(3_500L, store.gameProgress?.puzzleAnalytics?.get("encrypted_note")?.elapsedMs)
    }

    @Test fun recordWrongAttempt_savesReason() {
        val store = InMemoryGameProgressStore(); val viewModel = makeViewModel(store, MutableTimeProvider(1_000))
        viewModel.recordWrongAttempt("audio_record", "audioOrderIncorrect")
        val analytics = store.gameProgress?.puzzleAnalytics?.get("audio_record"); assertEquals(1, analytics?.wrongAttemptCount); assertEquals(1, analytics?.wrongReasonCounts?.get("audioOrderIncorrect"))
    }

    @Test fun requestHint_incrementsPuzzleHintViews() {
        val store = InMemoryGameProgressStore(); val viewModel = makeViewModel(store, MutableTimeProvider(1_000))
        viewModel.requestHint("commit_graph.1", "hint"); viewModel.requestHint("commit_graph.1", "hint")
        assertEquals(2, store.gameProgress?.puzzleAnalytics?.get("commit_graph")?.hintViewCount); assertEquals(1, store.gameProgress?.hintCount)
    }

    @Test fun recordPuzzleExit_savesBackNavigation() {
        val store = InMemoryGameProgressStore(); val time = MutableTimeProvider(1_000); val viewModel = makeViewModel(store, time)
        viewModel.startPuzzleSession("access_log"); time.current = 2_250; viewModel.recordPuzzleExit(PuzzleExitReason.backNavigation)
        val event = store.gameProgress?.puzzleAnalytics?.get("access_log")?.exitEvents?.single(); assertEquals(PuzzleExitReason.backNavigation, event?.reason); assertEquals(1_250L, event?.elapsedMsAtExit)
    }

    @Test fun submitPlayerFeedback_savesDifficultyAndComment() {
        val store = InMemoryGameProgressStore(); val viewModel = makeViewModel(store, MutableTimeProvider(1_000))
        assertTrue(viewModel.submitPlayerFeedback(4, "커밋 그래프가 어려웠어요")); assertEquals(4, store.gameProgress?.playerFeedback?.difficultyRating); assertNotNull(store.gameProgress?.playerFeedback)
    }

    @Test fun denyAnalyticsConsent_doesNotRecordPuzzleAnalytics() {
        val store = InMemoryGameProgressStore(); val viewModel = makeViewModel(store, MutableTimeProvider(1_000), consent = false)
        viewModel.denyAnalyticsConsent(); viewModel.startPuzzleSession("encrypted_note"); viewModel.recordWrongAttempt("encrypted_note", "noteWordOrderIncorrect")
        assertTrue(store.gameProgress?.puzzleAnalytics?.isEmpty() == true)
    }

    @Test fun submitPlayerFeedback_enqueuesFinalJson() {
        val store = InMemoryGameProgressStore(); val viewModel = makeViewModel(store, MutableTimeProvider(1_000))
        viewModel.submitPlayerFeedback(3, "좋아요")
        assertTrue(store.gameProgress?.pendingAnalyticsUploads?.last()?.envelope?.isFinal == true)
    }
}

File: ios/EscapePhoneTests/PlaytestAnalyticsTests.swift
import XCTest
@testable import EscapePhone

@MainActor
final class PlaytestAnalyticsTests: XCTestCase {
    private final class MutableDateProvider: DateProviding { var now: Date; init(_ now: Date) { self.now = now } }
    private func makeApp(store: InMemoryGameProgressStore, time: MutableDateProvider, consent: Bool = true) -> AppContainer { let app = AppContainer(store: store, motion: MockMotionController(), haptics: NoOpHapticProvider(), ads: NoOpAdGateway(), device: NoOpPuzzleDeviceConnector(), dateProvider: time); if consent { app.grantAnalyticsConsent() }; return app }

    func test_completePuzzleSession_savesElapsedTime() {
        let store = InMemoryGameProgressStore(); let time = MutableDateProvider(Date(timeIntervalSince1970: 1)); let app = makeApp(store: store, time: time)
        app.startPuzzleSession("encrypted_note"); time.now = Date(timeIntervalSince1970: 4.5); app.completePuzzleSession("encrypted_note"); app.save()
        XCTAssertEqual(store.gameProgress?.puzzleAnalytics["encrypted_note"]?.elapsedMs, 3_500)
    }

    func test_recordWrongAttempt_savesReason() {
        let store = InMemoryGameProgressStore(); let app = makeApp(store: store, time: MutableDateProvider(Date(timeIntervalSince1970: 1)))
        app.recordWrongAttempt("audio_record", reason: "audioOrderIncorrect")
        XCTAssertEqual(store.gameProgress?.puzzleAnalytics["audio_record"]?.wrongAttemptCount, 1); XCTAssertEqual(store.gameProgress?.puzzleAnalytics["audio_record"]?.wrongReasonCounts["audioOrderIncorrect"], 1)
    }

    func test_requestHint_incrementsPuzzleHintViews() {
        let store = InMemoryGameProgressStore(); let app = makeApp(store: store, time: MutableDateProvider(Date(timeIntervalSince1970: 1)))
        app.requestHint("commit_graph.1"); app.requestHint("commit_graph.1")
        XCTAssertEqual(store.gameProgress?.puzzleAnalytics["commit_graph"]?.hintViewCount, 2); XCTAssertEqual(store.gameProgress?.hintCount, 1)
    }

    func test_recordPuzzleExit_savesBackNavigation() {
        let store = InMemoryGameProgressStore(); let time = MutableDateProvider(Date(timeIntervalSince1970: 1)); let app = makeApp(store: store, time: time)
        app.startPuzzleSession("access_log"); time.now = Date(timeIntervalSince1970: 2.25); app.recordPuzzleExit(.backNavigation)
        XCTAssertEqual(store.gameProgress?.puzzleAnalytics["access_log"]?.exitEvents.first?.reason, .backNavigation); XCTAssertEqual(store.gameProgress?.puzzleAnalytics["access_log"]?.exitEvents.first?.elapsedMsAtExit, 1_250)
    }

    func test_submitPlayerFeedback_savesDifficultyAndComment() {
        let store = InMemoryGameProgressStore(); let app = makeApp(store: store, time: MutableDateProvider(Date(timeIntervalSince1970: 1)))
        XCTAssertTrue(app.submitPlayerFeedback(4, comment: "커밋 그래프가 어려웠어요")); XCTAssertEqual(store.gameProgress?.playerFeedback?.difficultyRating, 4); XCTAssertNotNil(store.gameProgress?.playerFeedback)
    }


    func test_denyAnalyticsConsent_doesNotRecordPuzzleAnalytics() {
        let store = InMemoryGameProgressStore(); let app = makeApp(store: store, time: MutableDateProvider(Date(timeIntervalSince1970: 1)), consent: false)
        app.denyAnalyticsConsent(); app.startPuzzleSession("encrypted_note"); app.recordWrongAttempt("encrypted_note", reason: "noteWordOrderIncorrect")
        XCTAssertTrue(store.gameProgress?.puzzleAnalytics.isEmpty == true)
    }

    func test_submitPlayerFeedback_enqueuesFinalJson() {
        let store = InMemoryGameProgressStore(); let app = makeApp(store: store, time: MutableDateProvider(Date(timeIntervalSince1970: 1)))
        app.submitPlayerFeedback(3, comment: "좋아요")
        XCTAssertEqual(store.gameProgress?.pendingAnalyticsUploads.last?.envelope.isFinal, true)
    }
}
